### ISSUE TITLE
Investigate server shutdown semantics in Cython code and possibly below

### ISSUE THREAD

**Issue by nathanielmanistaatgoogle** [grpc/grpc#6597]
`grpc_server_shutdown_and_notify` reads ["Send a GRPC_OP_COMPLETE event when there are no more calls being serviced."](https://github.com/grpc/grpc/blob/1ba1bba66a18b6b7986a1cfa52c6f1ac4a14a029/include/grpc/grpc.h#L349)

Does [this test](https://github.com/nathanielmanistaatgoogle/grpc/blob/fc5795fd2d66f1d162146731d3f1d52f47f581b0/src/python/grpcio/tests/unit/_needs_investimagation_test.py) show a violation of that guarantee in its flaky failures (it flakes around 2.5% for me)? Or is there something wrong in the test, or in our Cython code?

The gRPC Python beta server [keeps track of active RPCs and which events are due](https://github.com/grpc/grpc/blob/1ba1bba66a18b6b7986a1cfa52c6f1ac4a14a029/src/python/grpcio/grpc/_links/service.py#L296); I'm trying to shed that bookkeeping and rely on semantics of "[the server thread should just keep polling the server completion queue until it sees the shutdown tag, at which point it should break out of its loop and complete](https://github.com/nathanielmanistaatgoogle/grpc/blob/40a201e11255a4546d368197bba1db41a7fa5da3/src/python/grpcio/grpc/_server.py#L517)". That's supposed to work, right?

**Comment by ctiller** [comment]
At a guess (still grokking your code): whilst the server might know that no
calls are being serviced, it may manage to enqueue the shutdown completion
before the call completions (no guarantee is made of the ordering these
events will be seen from the completion queue).

**Comment by ctiller** [comment]
Shutdown completion mentioned above is the server shutdown notification
tag, not the completion queue shutdown event. The completion queue shutdown
event IS guaranteed to be the last thing ever produced from a completion
queue.

**Comment by nathanielmanistaatgoogle** [comment]
@ctiller, @soltanmm: [this test](https://github.com/nathanielmanistaatgoogle/grpc/blob/5ac93ce2d31584dcf8a1141d31483f452af1a239/src/python/grpcio/tests/unit/_really_test.py) is a stripped-down version that fails deterministically (in my development environments) and I hope should be readable. As per usual: what am I missing? And if I'm not missing anything, what's the defect in `cygrpc`? Is the test the simplest possible expression of "RPC is active, server-side application is taking a really really long time to decide what initial metadata to send for the RPC, while it is taking that time the server operator decides to shut down the server in a hurry"? If so, is this a scenario that is tested in other languages or in the Core's tests?

**Comment by soltanmm** [comment]
W.r.t. the linked (and 'thank-you-so-much-for-distilling-it'd) code, [line 64](https://github.com/nathanielmanistaatgoogle/grpc/blob/5ac93ce2d31584dcf8a1141d31483f452af1a239/src/python/grpcio/tests/unit/_really_test.py#L64) assumes that the server when shutting down is the last event coming out of the completion queue, which it might not be.

Due to the mishmash of SSH'ing and some degree of laziness (sorry), you're gonna get a patch as opposed to an actual file to look at:

```
62,64d61
<         if event.type == cygrpc.CompletionType.queue_shutdown:
<           break
<         condition.notify_all()
65a63,65
>         condition.notify_all()
>         if event.tag is _SERVER_SHUTDOWN_TAG:
>           break
126d125
<     server_completion_queue.shutdown()
```

That makes the test pass.

**Comment by soltanmm** [comment]
... I might have made the patch in reverse. I can never remember which direction one's supposed to write the `diff` command.

**Comment by nathanielmanistaatgoogle** [comment]
@soltanmm: well that's exactly the thought process through which we went on Tuesday afternoon. Either there needs to be a guarantee that `_SERVER_SHUTDOWN_TAG` will not be produced from the completion queue until after all calls have been destroyed or application threads processing calls need to be signaled in some way that a call to cancel_all_calls_and_notify has taken place and that they are no longer safe to start send_message and receive_message operations.

@ctiller: it seemed like over the course of our discussion the other day you reasoned your way through to guaranteeing that the server shutdown event won't be published on the completion queue until after all calls have been destroyed - does [this test](https://github.com/nathanielmanistaatgoogle/grpc/blob/5ac93ce2d31584dcf8a1141d31483f452af1a239/src/python/grpcio/tests/unit/_really_test.py) show a violation of that guarantee?

**Comment by soltanmm** [comment]
While I wouldn't put it past me that I'd gone hunger-catatonic at some key point of that conversation and/or code browsing exchange, I could've sworn we'd determined the ordered shutdown tag semantics were not guaranteed across arbitrary transports for... reasons (performance?)... and thus not something that (even if the ordering was present *today*) we wanted to rely on.

**Comment by ctiller** [comment]
@nathanielmanistaatgoogle I'm still working through the test, but one case I missed thinking through when we last spoke:

Requested calls need to be failed by the server during shutdown, and it's currently not guaranteed that those failures will happen before the server shutdown event.

**Comment by nathanielmanistaatgoogle** [comment]
Reprioritized and remilestoned - we've worked around this for now; it's no longer a must-answer-and-act-before-GA.

**Comment by muxi** [comment]
@nathanielmanistaatgoogle is this resolved?

**Comment by nathanielmanistaatgoogle** [comment]
@muxi: no, this is still at least a little interesting and something to be resolved in the next few months (1.9? 1.10?).
